Xrfi partial io

anguta opened the pull request:
allows for `xrfi_run` to compute metrics in chunks of baselines with new `Nwf_per_load` arg. Also modularized  and streamlined `xrfi_run` code, eliminating lots of copy-pasted stuff. Addresses #330  #346 

Potential issue: There is not yet a unit test for whether or not the reorganized xrfi_run produces identical results to the old xrfi_run. xrfi detection seems to be a very heuristic pursuit so I wasn't sure I should dwell on this but I'm open to adding it and would welcome feedback on the right way to do it (add back the old xrfi run for a few more merges until we delete it?)

anguta commented:
> allows for `xrfi_run` to compute metrics in chunks of baselines with new `Nwf_per_load` arg. Also modularized and streamlined `xrfi_run` code, eliminating lots of copy-pasted stuff.
> 
> Potential issue: There is not yet a unit test for whether or not the reorganized xrfi_run produces identical results to the old xrfi_run. xrfi detection seems to be a very heuristic pursuit so I wasn't sure I should dwell on this but I'm open to adding it and would welcome feedback on the right way to do it (add back the old xrfi run for a few more merges until we delete it?)



> allows for `xrfi_run` to compute metrics in chunks of baselines with new `Nwf_per_load` arg. Also modularized and streamlined `xrfi_run` code, eliminating lots of copy-pasted stuff. Addresses #330 #346
> 
> Potential issue: There is not yet a unit test for whether or not the reorganized xrfi_run produces identical results to the old xrfi_run. xrfi detection seems to be a very heuristic pursuit so I wasn't sure I should dwell on this but I'm open to adding it and would welcome feedback on the right way to do it (add back the old xrfi run for a few more merges until we delete it?)

I have now compared this branch with the old xrfi. I uncovered significant bug that stemmed from my misreading of `uvdata.get_antpairpols()` and now the old and new versions agree. I have tested agreement on the cases where all data products are provided and (a) crosscorr median filter is not computed and (b) a crosscorr median filter is computed. I think this is now ready for a close look.

jsdillon commented:
Ok, the other PR is merged in. LMK when all merge conflicts are fixed and this is ready for review.

anguta commented on hera_qm/xrfi.py:
```diff
@@ -1587,409 +1872,254 @@ def xrfi_run(ocalfits_files=None, acalfits_files=None, model_files=None, data_fi
+    # vdict stores all of the inputs and outputs (metrics, flags etc...) All possible products
+    # are ultimately referenced by vdict but could be set to None. outputs set to None are simply
+    # not written and can be used to determine whether steps depending on them should be run.
+    vdict = {'ocalfits_files': ocalfits_files, 'acalfits_files': acalfits_files,
```
they can be split up.

anguta commented on hera_qm/xrfi.py:
```diff
@@ -1587,409 +1872,254 @@ def xrfi_run(ocalfits_files=None, acalfits_files=None, model_files=None, data_fi
-        uvf_da = None; uvf_daf = None
-
-    # It is possible to have no median filters run so metrics could have len 0.
+    metrics = [] # this list stores metrics for median filters
```
that makes sense. I've split this up.

anguta commented on hera_qm/xrfi.py:
```diff
@@ -1587,409 +1872,254 @@ def xrfi_run(ocalfits_files=None, acalfits_files=None, model_files=None, data_fi
+                                                                                     run_check_acceptability=run_check_acceptability)
+    # now do omnivis and data files.
+    # add the names of the model and data files to vdict.
+    vdict['model_files'] = model_files; vdict['data_files'] = data_files
```
yep. I've removed this line.

anguta commented on hera_qm/xrfi.py:
```diff
@@ -1587,409 +1872,254 @@ def xrfi_run(ocalfits_files=None, acalfits_files=None, model_files=None, data_fi
-            flags += [uvf_oxf2]
+    # medfilt -> meanfilt.
+    cal_algs = ['detrend_meanfilt', 'detrend_meanfilt', 'zscore_full_array']
+    labels = [' gains, mean filter.', ' chisq, mean filter.', ' Renormalized chisq, median filter, round 2.']
```
I agree with changing this to overall z-score.

anguta commented on hera_qm/xrfi.py:
```diff
@@ -1587,409 +1872,254 @@ def xrfi_run(ocalfits_files=None, acalfits_files=None, model_files=None, data_fi
+    filter_switches = [omnivis_mean_filter, cross_mean_filter, auto_mean_filter]
+    algs = ['detrend_meanfilt', 'detrend_meanfilt', 'detrend_meanfilt']
+    for mf, ff, switch, alg, label, corr, input, uv, input, api in zip(uvmetrics, uvflags, filter_switches, algs, labels, correlations, input_uvs, uvs, input_uvs, apply_inits):
+            vdict[uv], vdict[mf], vdict[ff], _, metrics, flags = xrfi_run_step(uv=vdict[uv], uv_files=vdict[input], alg=alg, kt_size=kt_size, kf_size=kf_size,
```
fixed the overindents. For general pep8 issues, I'm not really an expert and not sure which specific pep8 rules we want to adapt/ignore so I think a repository level standard (pycodstyle check with some subset disabled?) is in order (future PR).

anguta commented on hera_qm/xrfi.py:
```diff
@@ -1288,35 +1280,42 @@ def xrfi_pipe(uv, alg='detrend_medfilt', Kt=8, Kf=8, xants=[], cal_mode='gain',
-                             run_check=run_check, check_extra=check_extra,
-                             run_check_acceptability=run_check_acceptability)
-    uvf_fws.label += ' Flags.'
+    if not isinstance(uv, UVFlag):
```
done this.

anguta commented on hera_qm/xrfi.py:
```diff
@@ -1381,6 +1380,287 @@ def chi_sq_pipe(uv, alg='zscore_full_array', modified=False, sig_init=6.0,
+                uv = UVCal()
+                # No partial i/o for uvcal yet.
+                uv.read_calfits(uv_files)
+            elif dtype=='uvdata':
```
Now raises an error.

anguta commented on hera_qm/xrfi.py:
```diff
@@ -1381,6 +1380,287 @@ def chi_sq_pipe(uv, alg='zscore_full_array', modified=False, sig_init=6.0,
+        # if we want, we can reinitialize uv
+        if reinitialize:
+            if uv_files is not None:
+                if issubclass(uv.__class__, UVData):
```
I don't really know if one is better then the other.

anguta commented on hera_qm/xrfi.py:
```diff
@@ -1381,6 +1380,287 @@ def chi_sq_pipe(uv, alg='zscore_full_array', modified=False, sig_init=6.0,
+            if run_filter:
+                # if run_filter is true, perform chi_sq_pipe or xrfi_pipe
+                if alg in ['zscore_full_array']:
+                    uvf, uvf_f = chi_sq_pipe(uv, alg=alg, modified=True, sig_init=sig_init, sig_adj=sig_adj,
```
I can create an issue. Not sure what is meant by before and after this PR since from what I could tell, the previous versions of `xrfi_run` only computed modified overall z-scores.

anguta commented on hera_qm/xrfi.py:
```diff
@@ -1414,8 +1694,8 @@ def xrfi_run(ocalfits_files=None, acalfits_files=None, model_files=None, data_fi
 
-    User must provide at least one an ocalfits_file, acalfits_file, model_file,
-    or data_file.
+    User must provide at least one an ocalfits_files, acalfits_files, model_files,
```
fixed.

anguta commented:
Some comments on comparing old and new versions of `xrfi`. I ran the following code for both the old, and new `xrfi_run` calls.

```
tmp_path = './xrfi_comparison/old/zen.2457698.40355.xrfi/'
fake_obs = 'zen.2457698.40355.HH'
ocal_file = os.path.join(tmp_path, fake_obs + '.omni.calfits')
shutil.copyfile(test_c_file, ocal_file)
acal_file = os.path.join(tmp_path, fake_obs + '.abs.calfits')
shutil.copyfile(test_c_file, acal_file)
raw_dfile = os.path.join(tmp_path, fake_obs + '.uvh5')
shutil.copyfile(test_uvh5_file, raw_dfile)
model_file = os.path.join(tmp_path, fake_obs + '.omni_vis.uvh5')
shutil.copyfile(test_uvh5_file, model_file)
xrfi.xrfi_run(ocal_file, acal_file, model_file, raw_dfile, 'Just a test', kt_size=3, clobber=True)
```

Now the new version of `xrfi` (this PR)


```
 # setup files in tmpdir.
tmp_path = './xrfi_comparison/new/'
fake_obs = 'zen.2457698.40355.HH'
ocal_file = os.path.join(tmp_path, fake_obs + '.omni.calfits')
shutil.copyfile(test_c_file, ocal_file)
acal_file = os.path.join(tmp_path, fake_obs + '.abs.calfits')
shutil.copyfile(test_c_file, acal_file)
raw_dfile = os.path.join(tmp_path, fake_obs + '.uvh5')
shutil.copyfile(test_uvh5_file, raw_dfile)
model_file = os.path.join(tmp_path, fake_obs + '.omni_vis.uvh5')
shutil.copyfile(test_uvh5_file, model_file)
xrfi.xrfi_run(ocal_file, acal_file, model_file, raw_dfile, 'Just a test', kt_size=3, clobber=True)
```

And now do this version of the PR but also use the partial i/o feature

```
# setup files in tmpdir.
tmp_path = '/Users/aewallwi/Dropbox/Aaron/Science/HERA-Team/debugging/hera_qm/xrfi_partial_io/xrfi_comparison/partial_io/'
fake_obs = 'zen.2457698.40355.HH'
ocal_file = os.path.join(tmp_path, fake_obs + '.omni.calfits')
shutil.copyfile(test_c_file, ocal_file)
acal_file = os.path.join(tmp_path, fake_obs + '.abs.calfits')
shutil.copyfile(test_c_file, acal_file)
raw_dfile = os.path.join(tmp_path, fake_obs + '.uvh5')
shutil.copyfile(test_uvh5_file, raw_dfile)
model_file = os.path.join(tmp_path, fake_obs + '.omni_vis.uvh5')
shutil.copyfile(test_uvh5_file, model_file)
xrfi.xrfi_run(ocal_file, acal_file, model_file, raw_dfile, 'Just a test', kt_size=3, Nwf_per_load=1, clobber=True)
```


```
old_files = sorted(glob.glob('./xrfi_comparison/old/zen.2457698.40355.xrfi/*.h5'))
new_files = sorted(glob.glob('./xrfi_comparison/new/zen.2457698.40355.xrfi/*.h5'))
pio_files = sorted(glob.glob('./xrfi_comparison/partial_io/zen.2457698.40355.xrfi/*.h5'))

assert len(old_files) == len(new_files)
for f1, f2, f3 in zip(old_files, new_files, pio_files):
    assert f1.split('/')[-1] == f2.split('/')[-1]
    assert f3.split('/')[-1] == f2.split('/')[-1]
    
    uvf1 = UVFlag(f1)
    uvf2 = UVFlag(f2)
    uvf3 = UVFlag(f3)
    print(f1)
    if uvf1.mode == 'metric':
        assert np.all(np.isclose(uvf1.metric_array, uvf2.metric_array))
        assert np.all(np.isclose(uvf2.metric_array, uvf3.metric_array))
    elif uvf1.mode == 'flag':
        assert np.all(np.isclose(uvf2.flag_array, uvf3.flag_array))
        assert np.all(np.isclose(uvf1.flag_array, uvf2.flag_array))
```
For a visual comparison, here are the combined flags after second round. 

```

uvf1 = UVFlag( './xrfi_comparison/old/zen.2457698.40355.xrfi/zen.2457698.40355.HH.combined_flags2.h5')
uvf2 = UVFlag( './xrfi_comparison/new/zen.2457698.40355.xrfi/zen.2457698.40355.HH.combined_flags2.h5')
ext = [uvf1.freq_array.min() / 1e6, uvf1.freq_array.max() / 1e6, uvf1.time_array.max(), uvf1.time_array.min()]
fig,axarr = plt.subplots(1,2)
plt.sca(axarr[0])
plt.imshow(uvf1.flag_array.squeeze(), aspect='auto', extent=ext)
plt.title('old xrfi')
plt.sca(axarr[1])
plt.imshow(uvf2.flag_array.squeeze(), aspect='auto', extent=ext)
plt.title('new xrfi')
assert np.all(np.isclose(uvf1.flag_array, uvf2.flag_array))
plt.gcf().set_size_inches(12,10)
```

![image](https://user-images.githubusercontent.com/6164284/89153249-61354080-d519-11ea-8ac4-93e268a03671.png)

anguta commented:
Thanks for the review @jsdillon. I think this is ready for another look.

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -2008,7 +2014,7 @@ def xrfi_run(ocalfits_files=None, acalfits_files=None, model_files=None, data_fi
     # medfilt -> meanfilt.
     cal_algs = ['detrend_meanfilt', 'detrend_meanfilt', 'zscore_full_array']
-    labels = [' gains, mean filter.', ' chisq, mean filter.', ' Renormalized chisq, median filter, round 2.']
+    labels = [' gains, mean filter.', ' chisq, mean filter.', ' overall z-score, median filter, round 2.']
```
One minor comment: I think it's much clearer if you say that it's the overall z-score of abscal/omnical chi^2. Overall z-score is vague.

jsdillon replied:
I've gone and fixed this myself. If you're OK with my change, feel free to merge.

anguta replied:
I'm good with this change.

anguta commented on hera_qm/xrfi.py:
```diff
@@ -1587,409 +1872,254 @@ def xrfi_run(ocalfits_files=None, acalfits_files=None, model_files=None, data_fi
-            metrics += [uvf_ox2]
-            flags += [uvf_oxf2]
+    # medfilt -> meanfilt.
+    cal_algs = ['detrend_meanfilt', 'detrend_meanfilt', 'zscore_full_array']
```
Right now, both rounds have `modified=True`. This is the case for this code before this specific PR. However, I think that this accidentally got introduced in a previous commit and it should be `modified=False`. Here is an older version which shows that we were using `modified=False` in the second round. 

https://github.com/HERA-Team/hera_qm/blob/00e0901f477eaa8a37a9fd0a31b47a823ccacaef/hera_qm/xrfi.py#L1833
